### PR TITLE
[ATOM-16287] Validate Alignment Across All floatNxM Between vulkan & …

### PR DESCRIPTION
**Pull request by galibzon**
[ATOM-16287] Validate Alignment Across All floatNxM Between vulkan & dx12

Added a new Example named "RHI/MatrixAlignmentTest" that can be used
to spot SRG Constant Buffer alignment issues when a float<R>x<C> is
followed by a "float" or a "float2" variable.

It's been found that, although We try to enforce DX12 alignment rules
across all RHIs, We end up encountering cases where such requested
alignments don't match, and this ASV example can be used to detect
the broken cases.

Signed-off-by: garrieta <garrieta@amazon.com>

### CODE REVIEW

**Comment by galibzon** [comment]
A snapshot of how it looks:

image

File: Gem/Code/Source/RHI/MatrixAlignmentTestExampleComponent.cpp [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+        {
+            serializeContext->Class<MatrixAlignmentTestExampleComponent, AZ::Component>()
+                ->Version(0)
+                ;
+        }
+    }
+
+    void MatrixAlignmentTestExampleComponent::OnTick([[maybe_unused]] float deltaTime, AZ::ScriptTimePoint time)
+    {
+        AZ_UNUSED(time);
```
Comment (jiaweig-amzn) [inline Gem/Code/Source/RHI/MatrixAlignmentTestExampleComponent.cpp:35]: using [[maybe_unused]] instead?

File: Gem/Code/Source/RHI/BasicRHIComponent.cpp [hunk 2]
```diff
@@ -439,6 +439,39 @@ namespace AtomSampleViewer
         return shader;
     }
 
+    AZ::Data::Instance<AZ::RPI::Shader> BasicRHIComponent::LoadShader(const char* shaderFilePath, const char* sampleName, const AZ::Name& supervariantName)
```
[the remark this answers is not in the document]
Comment (galibzon) [reply Gem/Code/Source/RHI/BasicRHIComponent.cpp:442]: Thanks, wil consolidate the LoadShader calls.

File: Gem/Code/Source/RHI/MatrixAlignmentTestExampleComponent.cpp (hunk 1 above)
Comment (galibzon) [reply Gem/Code/Source/RHI/MatrixAlignmentTestExampleComponent.cpp:35]: You are right.

**Review by jiaweig-amzn — APPROVED** [review]

**Review by Adi-Amazon — APPROVED** [review]

File: Gem/Code/Source/RHI/MatrixAlignmentTestExampleComponent.cpp [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+            FindShaderInputIndex(&m_resolutionConstantIndex, m_shaderResourceGroup, resolutionConstantId, sampleName);
+
+            const Name rowSelectionConstantId{ "m_numRows" };
+            FindShaderInputIndex(&m_numRowsConstantIndex, m_shaderResourceGroup, rowSelectionConstantId, sampleName);
+
+            const Name colSelectionConstantId{ "m_numColumns" };
+            FindShaderInputIndex(&m_numColumnsConstantIndex, m_shaderResourceGroup, colSelectionConstantId, sampleName);
+
+            {
+                const Name MatrixName{ "m_matrix11" };
```
Comment (moudgils) [inline Gem/Code/Source/RHI/MatrixAlignmentTestExampleComponent.cpp:314]: Minor nit: Can we refactor these four lines to a method?

**Review by moudgils — APPROVED** [review]

**Pull request merged**